04-evm-mapping.md

```diff
@@ -0,0 +1,307 @@
+2. It exceeds the EVM's 256 bit architecture.
+
+Problem 1 renders Solidity smart contracts instantly incompatible with the Filecoin addressing scheme, as well as EVM opcodes that take or return addresses for arguments, e.g. CALLER, CALL, CALLCODE, DELEGATECALL, COINBASE, etc. This problem is hard to work around, and would require a fork of the EVM to modify existing opcodes for semantic awareness of addresses (although this is really hard to get right), or to introduce a Filecoin-specific opcode family to deal Filecoin addresses (e.g. FCALL, FCALLCODE, etc.) The latter would break as-is deployability of existing smart contracts.
+
+Problem 2 can be workable by spilling over and combining values in the stack, through Filecoin-specific Solidity libraries.
+
+Find the proposed solution next, and a summary of alternatives considered in Annex A.
+
+### Proposed solution: universal stable addresses
+
+The idea is to promote address class 2 into a unified fixed-size stable address space all actors, **including account actors**. At present, this class only identifies **non-account actors**. The only address space that covers all actors is the ID class, which is not reorg-stable and thus unsafe.
+
+The width of this address space is already 20 bytes. EVM actors would only interact with class 2 addresses, with the prefix stripped to honour the 160-bit address expectation, for full EVM bytecode compatibility.
+
+Nevertheless, it has been noted that the security of 20-byte addreses is beginning to be insufficient, with an estimated cost of $10bn to find an arbitrary collision by carrying out 2^80 computations.
+
+Therefore, the actual proposal is **to introduce a new address class 4 for a 256-bit stable address space**. This class would become the **canonical stable address class**, with class 2 addresses (20-byte) mapping over to their class 4 equivalents (32 bytes). Class 2 can thus be thought of as an alias/symlink to class 4, to be used in environments that are unable handle larger widths.
+
+The state tree would track class 2 and class 4 address for all actors by maintaining the necessary indices.
+
+**Class 2 and 4 address derivation**
+
+Class 2 and 4 addresses are protocol-derived, by hashing the relevant cryptographic or input material.
+
+With a high probability, the hash function of class 4 will remain BLAKE2, with a 256-bit digest size. Class 2 will continue relying on blake2b-160.
+
+1. For secp256k1 account actors (class 1), the preimage is the pubkey. The value of the class 2 address is identical to that of class 1.
+2. For BLS account actors (class 3), the preimage is the pubkey.
+3. For FVM native actors, the preimage is `sender || nonce || # of actors created during message execution`.
+4. For EVM foreign actors, the preimage is inherited from CREATE and CREATE2.
```

Comment:
Unfortunately, this would make the EVM "special". I'd prefer not to have "special" address construction methods for every foreign runtime.

I'd rather use the same algorithm _everywhere_ (EVM and FVM): `sender || target code cid || salt || init params`. This isn't identical to how the EVM currently works, but provides the same guarantees.

We'd need to do a bit of research on how predictable addresses are used, but I _believe_ they're mostly used by external tooling for some payment channel use-cases. If that's the case, it should be pretty trivial to drop-in a replacement function as long as that replacement function has the same inputs.